framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,563 @@
+        String nextComponentType =
+                arrayType.indexOf('[') == -1
+                        ? null
+                        : arrayType.substring(0, arrayType.lastIndexOf('['));
+        // base case when the component is a non-array type
+        if (nextComponentType == null) {
+            return formatAnnotations(e.tlAnnotationsHere) + arrayType + " " + result.toString();
+        } else {
+            result.append(formatAnnotations(e.tlAnnotationsHere));
+            result.append("[] ");
+        }
+        // find the next array type; this loop should always find something, so throws an
+        // exception afterward if it fails
+        ATypeElement innerType = null;
+        for (Map.Entry<InnerTypeLocation, ATypeElement> ite : e.innerTypes.entrySet()) {
+            InnerTypeLocation loc = ite.getKey();
+            ATypeElement it = ite.getValue();
+            if (loc.location.contains(TypePathEntry.ARRAY)) {
+                innerType = it;
+            }
+        }
+
+        if (innerType == null) {
+            throw new BugInCF(
+                    "Encountered a multidimensional array without an inner type. This probably indicates a bug in the Annotation File Utilities.");
+        }
+
+        return formatArrayTypeImpl(innerType, nextComponentType, result);
+    }
+
+    /**
+     * Formats an AField so that it can be printed in a stub. An AField represents a variable
+     * declaration. In practice, {@code aField} should represent either a field declaration or a
+     * formal parameter of a method, because stub files should not contain local variable
+     * declarations. It is the responsibility of each caller of this method to ensure that is true.
+     *
+     * <p>It is also the responsibility of the caller to place the output of this method in context;
+     * because it is shared between field declarations and formal parameters, it does not add any
+     * trailing semicolons/commas/other syntax.
+     *
+     * @param aField the field to format
+     * @param fieldName the name to use for the declaration in the stub file. This doesn't matter
+     *     for parameters, but must be correct for fields.
+     * @param className the simple name of the enclosing class. This is only used for printing the
+     *     type of an explicit receiver parameter (i.e., a parameter named "this").
+     * @return a String suitable to print in a stub file
+     */
+    private static String formatAField(AFieldWrapper aField, String fieldName, String className) {
+        StringBuilder result = new StringBuilder();
+        String basetype;
+        if ("this".equals(fieldName)) {
+            basetype = className;
+        } else {
+            basetype = aField.getType();
+        }
+
+        if (basetype.contains("[")) {
+            String formattedArrayType = formatArrayType(aField.getTheField().type, basetype);
+            result.append(formattedArrayType);
```

Comment:
This doesn't seem to have a trailing space, though the `else` clause does.  Please fix, or clarify documentation here or elsewhere.